### ISSUE TITLE
Suggestion: Allow interface to extend conditional type

### ISSUE THREAD

**Issue by infogulch** [microsoft/TypeScript#23850]
An interface cannot currently extend a conditional type. See: ([playground link](https://www.typescriptlang.org/play/index.html#src=type%20Test%3CT%3E%20%3D%20T%20extends%20%7B%7D%20%3F%20%7B%20test%3A%20T%20%7D%20%3A%20%7B%7D%0D%0A%0D%0A%2F%2F%20Ok%3A%20type%20with%20selected%20condition%0D%0Ainterface%20I%20extends%20Test%3Cnumber%3E%20%7B%20%7D%0D%0A%0D%0A%2F%2F%20%22Error%3A%20An%20interface%20may%20only%20extend%20a%20class%20or%20another%20interface.%22%22%0D%0Ainterface%20J%3CT%3E%20extends%20Test%3CT%3E%20%7B%20%7D%0D%0A%0D%0A%2F%2F%20Ok%3A%20type%20without%20condition%0D%0Atype%20Test2%3CT%3E%20%3D%20%7B%20test%3A%20T%20%7D%0D%0Ainterface%20K%3CT%3E%20extends%20Test2%3CT%3E%20%7B%20%7D))
```typescript
type Test<T> = T extends {} ? { test: T } : {}

// Ok: type with selected condition
interface I extends Test<number> { }

// "Error: An interface may only extend a class or another interface."
interface J<T> extends Test<T> { }

// Ok: type without condition
type Test2<T> = { test: T }
interface K<T> extends Test2<T> { }
```

This example exhibits one of these problems:

* `J<T>` should be allowed to extend `Test<T>`, to make it more complete and allow conditional type things in interfaces.
--*OR*--
* The error message should be made more clear, because interfaces can obviously extend types as you can see above.

**Comment by abhiravredox** [comment]
Hi, I would like to fix this bug; Please help me fix it.

**Comment by infogulch** [comment]
Could it enumerate all possible end types and check that it's valid to extend all of them?

**Comment by mhegazy** [comment]
> Could it enumerate all possible end types and check that it's valid to extend all of them?

No. the issue is that now your interface itself is a conditional type.. and based on the condition it extends one or the other. 

The compiler makes assumptions about interfaces and classes, being object types. and they can not suddenly be conditional, union or intersection.

**Comment by mhegazy** [comment]
We originally picked this error message to avoid having to go through all possible types, i.e. it can not be a union either, or an intersection that includes a generic type, or a mapped type with a generic type involved, or a an index access type if a generic type is involved.. etc..

**Comment by infogulch** [comment]
Hmm. If all the possible types the condition could evaluate to are object types, and an interface extends it, the interface is still an object type even if it's conditional, no? I.e. even though it's conditional, it's still guaranteed to resolve to some object type.

If that's not a strong enough guarantee because of compiler/typesystem/implementation reasons, I won't argue it any farther. Thank you for your explanations!

**Comment by MartinJohns** [comment]
@RyanCavanaugh 

> For example, you can write `interface I extends number {`

You can't. It results in this error:

> 'number' only refers to a type, but is being used as a value here.

**Comment by m-henderson** [comment]
@RyanCavanaugh you mentioned the only thing we can do is update the error message; @mhegazy do you agree with that? if so, what does the diagnostics message need to say so that it is accurate than the current diagnostics message?

**Comment by RyanCavanaugh** [comment]
The error message is improved as described
